arch/arm/core/cortex_a_r/swap_helper.S

```diff
@@ -24,14 +24,71 @@
 _ASM_FILE_PROLOGUE
 
 GTEXT(z_arm_svc)
+#ifdef CONFIG_USE_SWITCH
+GTEXT(z_arm_context_switch)
+#else
 GTEXT(z_arm_do_swap)
+#endif /* CONFIG_USE_SWITCH */
 GTEXT(z_do_kernel_oops)
 #if defined(CONFIG_USERSPACE)
 GTEXT(z_arm_do_syscall)
 #endif
 
 GDATA(_kernel)
 
+#ifdef CONFIG_USE_SWITCH
+/*
+ * Routine to handle context switches
+ *
+ * This function is directly called either by _isr_wrapper() in case of
+ * preemption, or arch_switch() in case of cooperative switching.
+ *
+ * void z_arm_context_switch(struct k_thread *new, struct k_thread *old);
+ */
+SECTION_FUNC(TEXT, z_arm_context_switch)
+
+	ldr r2, =_thread_offset_to_callee_saved
+	add r2, r1, r2
+
+	stm r2, {r4-r11, sp, lr}
```

Comment:
This is pleasingly simple, but FWIW making it a function call is an extra set of instructions that you really don't need in the cooperative switch case  It wouldn't be much more complicated to make arch_switch() an inline assembly expansion that does mostly the same stuff, adds a clobber for r0-3 and r12 [1], and stores the address of the following instruction into the thread context via a simple assembly symbol.

[1] Probably.  I'm winging my knowledge of the ARM ABI based on general intuition and a copy of the AAPCS spec I just googled in another window.  Apply salt.

Reply:
I'm reconsidering this now.